scrapy/downloadermiddlewares/retry.py

```diff
@@ -13,35 +13,27 @@
 """
 import logging
 
-from twisted.internet import defer
-from twisted.internet.error import TimeoutError, DNSLookupError, \
-        ConnectionRefusedError, ConnectionDone, ConnectError, \
-        ConnectionLost, TCPTimedOutError
-from twisted.web.client import ResponseFailed
-
+import six
 from scrapy.exceptions import NotConfigured
 from scrapy.utils.response import response_status_message
-from scrapy.core.downloader.handlers.http11 import TunnelError
 from scrapy.utils.python import global_object_name
+from scrapy.utils.misc import load_object
 
 logger = logging.getLogger(__name__)
 
 
 class RetryMiddleware(object):
 
-    # IOError is raised by the HttpCompression middleware when trying to
-    # decompress an empty response
-    EXCEPTIONS_TO_RETRY = (defer.TimeoutError, TimeoutError, DNSLookupError,
```

Comment:
Previously the way to add a new exception to retry was to subclass RetryMiddleware and add something to EXCEPTIONS_TO_RETRY. It means that the change is backwards incompatible. We should make sure that code like this still works, but shows a warning:

```py
from scrapy.downloadermiddlewares.retry import RetryMiddleware
from scrapy.core.downloader.handlers.http11 import TunnelError

class MyRetryMiddleware(RetryMiddleware):
    EXCEPTIONS_TO_RETRY = RetryMiddleware.EXCEPTIONS_TO_RETRY + (TunnelError,)
```

 I think we can make it backwards compatible this way:

1. make EXCEPTIONS_TO_RETRY a property; raise a warning if it is accessed. This property should return the same exception list. You can load it from default_settings if no settings are available.

2. If EXCEPTIONS_TO_RETRY attribute is overridden, middleware should use this attribute instead of exceptions specified in settings.py.